File: lib/core/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

class WebSocketService {
  WebSocketChannel? _channel;
  StreamController<Map<String, dynamic>>? _messageController;
  Timer? _reconnectTimer;
  bool _isConnected = false;
  String? _token;
  String? _userId;

  // Singleton pattern
  static final WebSocketService _instance = WebSocketService._internal();
  factory WebSocketService() => _instance;
  WebSocketService._internal();

  // Stream getter
  Stream<Map<String, dynamic>> get messageStream {
    _messageController ??= StreamController<Map<String, dynamic>>.broadcast();
    return _messageController!.stream;
  }

  // Bağlantı durumu
  bool get isConnected => _isConnected;

  // WebSocket bağlantısını başlat
  Future<void> connect(String token, String userId) async {
    if (_isConnected) return;

    _token = token;
    _userId = userId;

    try {
      // WebSocket URL'i (Docker container için)
      final wsUrl = 'ws://10.0.2.2:3000/ws?token=$token&userId=$userId';

      _channel = WebSocketChannel.connect(Uri.parse(wsUrl));
      _isConnected = true;

      print('🔌 WebSocket bağlantısı kuruldu');

      // Mesajları dinle
      _channel!.stream.listen(
        (message) {
          _handleMessage(message);
        },
        onError: (error) {
          print('❌ WebSocket hatası: $error');
          _isConnected = false;
          _scheduleReconnect();
        },
        onDone: () {
          print('🔌 WebSocket bağlantısı kapandı');
          _isConnected = false;
          _scheduleReconnect();
        },
      );

      // Bağlantı mesajı gönder
      _sendMessage({
        'type': 'connect',
        'userId': userId,
        'timestamp': DateTime.now().toIso8601String(),
      });
    } catch (e) {
      print('❌ WebSocket bağlantı hatası: $e');
      _isConnected = false;
      _scheduleReconnect();
    }
  }

  // Mesaj gönder
  void _sendMessage(Map<String, dynamic> message) {
    if (_channel != null && _isConnected) {
      try {
        _channel!.sink.add(jsonEncode(message));
      } catch (e) {
        print('❌ WebSocket mesaj gönderme hatası: $e');
      }
    }
  }

  // Gelen mesajları işle
  void _handleMessage(dynamic message) {
    try {
      if (message is String) {
        final data = jsonDecode(message) as Map<String, dynamic>;
        _messageController?.add(data);

        // Mesaj tipine göre işle
        switch (data['type']) {
          case 'invoice_status_update':
            print(
                '📊 Fatura durumu güncellendi: ${data['invoiceId']} -> ${data['status']}');
            break;
          case 'package_status_update':
            print(
                '📦 Paket durumu güncellendi: ${data['packageId']} -> ${data['status']}');
            break;
          case 'processing_progress':
            print('⚙️ İşlem ilerlemesi: ${data['progress']}%');
            break;
          default:
            print('📨 WebSocket mesajı: $data');
        }
      }
    } catch (e) {
      print('❌ WebSocket mesaj işleme hatası: $e');
    }
  }

  // Yeniden bağlanma zamanlaması
  void _scheduleReconnect() {
    if (_reconnectTimer != null) return;

    _reconnectTimer = Timer(const Duration(seconds: 5), () {
      _reconnectTimer = null;
      if (_token != null && _userId != null) {
        print('🔄 WebSocket yeniden bağlanmaya çalışılıyor...');
        connect(_token!, _userId!);
      }
    });
  }

  // Belirli bir fatura için durum güncellemelerini dinle
  void listenToInvoice(String invoiceId) {
    _sendMessage({
      'type': 'listen_invoice',
      'invoiceId': invoiceId,
      'timestamp': DateTime.now().toIso8601String(),
    });
  }

  // Belirli bir paket için durum güncellemelerini dinle
  void listenToPackage(String packageId) {
    _sendMessage({
      'type': 'listen_package',
      'packageId': packageId,
      'timestamp': DateTime.now().toIso8601String(),
    });
  }

  // Tüm faturaları dinle
  void listenToAllInvoices() {
    _sendMessage({
      'type': 'listen_all_invoices',
      'timestamp': DateTime.now().toIso8601String(),
    });
  }

  // Bağlantıyı kapat
  void disconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;

    if (_channel != null) {
      _channel!.sink.close(status.goingAway);
      _channel = null;
    }

    _isConnected = false;
    print('🔌 WebSocket bağlantısı kapatıldı');
  }

  // Servisi temizle
  void dispose() {
    disconnect();
    _messageController?.close();
    _messageController = null;
  }
}

File: fatura_backend/src/services/websocket.service.js
const WebSocket = require('ws');
const jwt = require('jsonwebtoken');
require('dotenv').config(); // .env dosyasını yükle

class WebSocketService {
    constructor() {
        this.wss = null;
        this.clients = new Map(); // userId -> WebSocket bağlantıları
        this.authenticatedClients = new Map(); // token -> { userId, ws }
    }

    // WebSocket sunucusunu başlat
    initialize(server) {
        this.wss = new WebSocket.Server({ 
            server,
            path: '/ws'
        });

        this.wss.on('connection', (ws, req) => {
            this.handleConnection(ws, req);
        });

        console.log('🔌 WebSocket sunucusu başlatıldı');
    }

    // Yeni bağlantıyı işle
    handleConnection(ws, req) {
        console.log('🔌 Yeni WebSocket bağlantısı');

        // URL'den token ve userId'yi al
        const url = new URL(req.url, 'http://localhost');
        const token = url.searchParams.get('token');
        const userId = url.searchParams.get('userId');

        if (!token || !userId) {
            console.log('❌ WebSocket: Token veya userId eksik');
            ws.close(1008, 'Token veya userId eksik');
            return;
        }

        // Token'ı doğrula
        try {
            const decoded = jwt.verify(token, process.env.JWT_SECRET);
            if (decoded.uid !== userId) {
                console.log('❌ WebSocket: Token userId uyuşmazlığı');
                ws.close(1008, 'Token userId uyuşmazlığı');
                return;
            }

            // Bağlantıyı kaydet
            this.authenticatedClients.set(token, { userId, ws });
            
            // Kullanıcının bağlantılarını kaydet
            if (!this.clients.has(userId)) {
                this.clients.set(userId, new Set());
            }
            this.clients.get(userId).add(ws);

            console.log(`✅ WebSocket: Kullanıcı ${userId} bağlandı`);

            // Bağlantı mesajını gönder
            this.sendToClient(ws, {
                type: 'connected',
                userId: userId,
                timestamp: new Date().toISOString()
            });

        } catch (error) {
            console.log('❌ WebSocket: Token doğrulama hatası', error.message);
            ws.close(1008, 'Token doğrulama hatası');
            return;
        }

        // Mesaj dinle
        ws.on('message', (message) => {
            this.handleMessage(ws, message, userId);
        });

        // Bağlantı kapanma
        ws.on('close', () => {
            this.handleDisconnection(ws, userId, token);
        });

        // Hata durumu
        ws.on('error', (error) => {
            console.log('❌ WebSocket hatası:', error.message);
            this.handleDisconnection(ws, userId, token);
        });
    }

    // Gelen mesajı işle
    handleMessage(ws, message, userId) {
        try {
            const data = JSON.parse(message);
            console.log(`📨 WebSocket mesajı (${userId}):`, data.type);

            switch (data.type) {
                case 'listen_invoice':
                    this.handleListenInvoice(ws, data.invoiceId, userId);
                    break;
                case 'listen_package':
                    this.handleListenPackage(ws, data.packageId, userId);
                    break;
                case 'listen_all_invoices':
                    this.handleListenAllInvoices(ws, userId);
                    break;
                default:
                    console.log('❌ Bilinmeyen WebSocket mesaj tipi:', data.type);
            }
        } catch (error) {
            console.log('❌ WebSocket mesaj işleme hatası:', error.message);
        }
    }

    // Fatura dinleme
    handleListenInvoice(ws, invoiceId, userId) {
        console.log(`👂 Kullanıcı ${userId} fatura ${invoiceId} dinliyor`);
        // Burada fatura durumu değişikliklerini dinleyebiliriz
    }

    // Paket dinleme
    handleListenPackage(ws, packageId, userId) {
        console.log(`👂 Kullanıcı ${userId} paket ${packageId} dinliyor`);
        // Burada paket durumu değişikliklerini dinleyebiliriz
    }

    // Tüm faturaları dinleme
    handleListenAllInvoices(ws, userId) {
        console.log(`👂 Kullanıcı ${userId} tüm faturaları dinliyor`);
        // Burada tüm fatura durumu değişikliklerini dinleyebiliriz
    }

    // Bağlantı kapanma
    handleDisconnection(ws, userId, token) {
        console.log(`🔌 WebSocket: Kullanıcı ${userId} bağlantısı kapandı`);

        // Kullanıcının bağlantılarını temizle
        if (this.clients.has(userId)) {
            this.clients.get(userId).delete(ws);
            if (this.clients.get(userId).size === 0) {
                this.clients.delete(userId);
            }
        }

        // Authenticated clients'tan kaldır
        this.authenticatedClients.delete(token);
    }

    // Belirli bir client'a mesaj gönder
    sendToClient(ws, message) {
        if (ws.readyState === WebSocket.OPEN) {
            try {
                ws.send(JSON.stringify(message));
            } catch (error) {
                console.log('❌ WebSocket mesaj gönderme hatası:', error.message);
            }
        }
    }

    // Belirli bir kullanıcıya mesaj gönder
    sendToUser(userId, message) {
        if (this.clients.has(userId)) {
            const userConnections = this.clients.get(userId);
            userConnections.forEach(ws => {
                this.sendToClient(ws, message);
            });
        }
    }

    // Tüm kullanıcılara mesaj gönder
    broadcast(message) {
        this.wss.clients.forEach(client => {
            if (client.readyState === WebSocket.OPEN) {
                this.sendToClient(client, message);
            }
        });
    }

    // Fatura durumu güncellemesi gönder
    sendInvoiceStatusUpdate(userId, invoiceId, status, packageId = null) {
        const message = {
            type: 'invoice_status_update',
            invoiceId: invoiceId,
            status: status,
            packageId: packageId,
            timestamp: new Date().toISOString()
        };

        this.sendToUser(userId, message);
        console.log(`📊 Fatura durumu güncellendi: ${invoiceId} -> ${status}`);
    }

    // Paket durumu güncellemesi gönder
    sendPackageStatusUpdate(userId, packageId, status) {
        const message = {
            type: 'package_status_update',
            packageId: packageId,
            status: status,
            timestamp: new Date().toISOString()
        };

        this.sendToUser(userId, message);
        console.log(`📦 Paket durumu güncellendi: ${packageId} -> ${status}`);
    }

    // İşlem ilerlemesi gönder
    sendProcessingProgress(userId, progress, packageId = null) {
        const message = {
            type: 'processing_progress',
            progress: progress,
            packageId: packageId,
            timestamp: new Date().toISOString()
        };

        this.sendToUser(userId, message);
        console.log(`⚙️ İşlem ilerlemesi: %${progress}`);
    }

    // Sunucuyu kapat
    close() {
        if (this.wss) {
            this.wss.close();
            console.log('🔌 WebSocket sunucusu kapatıldı');
        }
    }
}

module.exports = new WebSocketService();
